src/js/common/utils/accessible-menus.js

```diff
@@ -0,0 +1,349 @@
+ *                               { menu: <HTMLElement>, menuButton: <HTMLElement> }
+ *                               Otherwise, returns null
+ */
+function getMenuStructure(menuLi) {
+  const menuButton = menuLi.querySelector('button, [role="button"]');
+  const menu = menuButton ? menuLi.querySelector(`#${menuButton.getAttribute('aria-controls')}`) : null;
+
+  if (!menuButton || !menu) {
+    return null;
+  }
+  const menuRole = (menu.getAttribute('role') || '').toLowerCase();
+  if (!(menuRole === 'menu' || menu.querySelector('[role="menu"]'))) {
+    return null;
+  }
+
+  return { menuButton, menu };
+}
+
+
+/**
+ * Closes a menu.
+ *
+ * @param {HTMLLIElement} menuLI  The <li> containing the menubutton and menu
+ */
+function closeMenu(menuLi) {
+  // If we're not dealing with a menu structure, abort
+  const struct = getMenuStructure(menuLi);
+  if (!struct) {
+    return;
+  }
+
+  const { menuButton, menu } = struct;
+
+  // Close the menu
+  menuButton.removeAttribute('aria-expanded');
+  menu.setAttribute('hidden', 'hidden');
+}
+
+
+/**
+ * Opens a menu.
+ *
+ * @param {HTMLLIElement} menuLi  The <li> containing the menubutton and menu
+ * @param {bool} openSubMenu      Whether or not the first submenu should be opened. E.g. if we're opening the
+ *                                 second-level menu in a wide screen, we want the third-level opened as well
+ * @param {bool} stealFocus       Whether or not to focus on the first item in the opened menu
+ */
+function openMenu(menuLi, openSubMenu = false, stealFocus = true) {
+  // If we're not dealing with a menu structure, abort
+  const struct = getMenuStructure(menuLi);
+  if (!struct) {
+    return;
+  }
+
+  const { menuButton, menu } = struct;
+
+  // First, close all sibling menus
+  const openMenus = menuLi.parentElement.querySelectorAll('[aria-expanded=true]');
+  openMenus.forEach(m => closeMenu(m.parentElement));
+
+  // Open the menu
+  menuButton.setAttribute('aria-expanded', true);
+  menu.removeAttribute('hidden');
+
+  if (stealFocus) {
+    const menuRole = (menu.getAttribute('role') || '').toLowerCase();
+    const firstMenuItem = menuRole === 'menu' ?
+      firstVisibleChild(menu) :
+      firstVisibleChild(menu.querySelector('[role="menu"]'));
+    firstMenuItem.firstElementChild.focus();
+  }
+
+  // If we're wide-screen (for example), open the first submenu
+  if (openSubMenu) {
+    openMenu(menu.firstElementChild, false, false);
+  }
+}
+
+
+/**
+ * Creates a closure which closes all menus in the provided element.
+ *
+ * @param {HTMLElement}
+ * @return {function}   Call this to close all menus
+ */
+function closeAll(menuElement) {
+  // Get all the parent <li>s of open menus
+  const allMenus = Array.from(menuElement.querySelectorAll('[aria-expanded=true]'))
+    .map(e => findNearestLi(e));
+
+  // Add the top-level 
+  if (menuElement.getAttribute('aria-expanded') === true) {
+    allMenus.push(findNearestLi(menuElement));
+  }
```

Comment:
Interesting. Why isn't this found with the `allMenus` query selector?

Reply:
Because `menuElement.querySelector()` won't select `menuElement`, so I have to look for it specifically. Alternatively, I could look at `menuElement`'s parent, but then it'd also look at any of `menuElement`'s siblings and could have unexpected side-effects, so I decided to just do it this way. :man_shrugging: